editor/editor_translation_parser.cpp

```diff
@@ -0,0 +1,167 @@
+	ClassDB::add_virtual_method(get_class_static(), MethodInfo(Variant::NIL, "parse_text", PropertyInfo(Variant::STRING, "path"), PropertyInfo(Variant::ARRAY, "extracted_strings")));
+	ClassDB::add_virtual_method(get_class_static(), MethodInfo(Variant::ARRAY, "get_recognized_extensions"));
+}
+
+/////////////////////////
+
+void EditorTranslationParser::get_recognized_extensions(List<String> *r_extensions) const {
+	Set<String> extensions;
+	List<String> temp;
+	for (int i = 0; i < standard_parsers.size(); i++) {
+		standard_parsers[i]->get_recognized_extensions(&temp);
+	}
+	for (int i = 0; i < custom_parsers.size(); i++) {
+		custom_parsers[i]->get_recognized_extensions(&temp);
+	}
+	// Remove duplicates.
+	for (int i = 0; i < temp.size(); i++) {
+		extensions.insert(temp[i]);
+	}
+	for (auto E = extensions.front(); E; E = E->next()) {
+		r_extensions->push_back(E->get());
+	}
+}
+
+bool EditorTranslationParser::can_parse(const String &p_extension) const {
+	List<String> extensions;
+	get_recognized_extensions(&extensions);
+	for (int i = 0; i < extensions.size(); i++) {
+		if (p_extension == extensions[i]) {
+			return true;
+		}
+	}
+	return false;
+}
+
+Ref<EditorTranslationParserPlugin> EditorTranslationParser::get_parser(const String &p_extension) const {
+	// Consider user-defined parsers first.
+	for (int i = 0; i < custom_parsers.size(); i++) {
+		List<String> temp;
+		custom_parsers[i]->get_recognized_extensions(&temp);
+		for (int j = 0; j < temp.size(); j++) {
+			if (temp[j] == p_extension) {
+				return custom_parsers[i];
+			}
+		}
+	}
+
+	for (int i = 0; i < standard_parsers.size(); i++) {
+		List<String> temp;
+		standard_parsers[i]->get_recognized_extensions(&temp);
+		for (int j = 0; j < temp.size(); j++) {
+			if (temp[j] == p_extension) {
+				return standard_parsers[i];
+			}
+		}
+	}
+
+	WARN_PRINT("No translation parser available for \"" + p_extension + "\" extension.");
+
+	return nullptr;
+}
+
+void EditorTranslationParser::add_parser(const Ref<EditorTranslationParserPlugin> &p_parser, ParserType p_type) {
+	if (p_type == ParserType::STANDARD) {
+		standard_parsers.push_back(p_parser);
+	} else if (p_type == ParserType::CUSTOM) {
+		custom_parsers.push_back(p_parser);
+	}
+}
+
+void EditorTranslationParser::remove_parser(const Ref<EditorTranslationParserPlugin> &p_parser, ParserType p_type) {
+	if (p_type == ParserType::STANDARD) {
+		standard_parsers.erase(p_parser);
+	} else if (p_type == ParserType::CUSTOM) {
+		custom_parsers.erase(p_parser);
+	}
+}
+
+EditorTranslationParser *EditorTranslationParser::get_singleton() {
+	if (!singleton) {
+		singleton = memnew(EditorTranslationParser);
+	}
+	return singleton;
+}
+
+EditorTranslationParser::EditorTranslationParser() {
+	//I'm not sure where to add these instances. Thought of EditorNode
+	Ref<EditorTranslationParserPluginGDScript> gdscript_plugin;
+	gdscript_plugin.instance();
+	add_parser(gdscript_plugin, ParserType::STANDARD);
```

Comment:
This should be done in `modules/gdscript`, the `editor/*` code should not have dependencies (includes) on modules.

You can do it in `modules/gdscript/register_types.cpp::register_gdscript_types()`, so when GDScript is registered, it will also register this translation parser plugin.

Reply:
Currently the added dependency on the GDScript module causes a build issue:
```
editor/libeditor.linuxbsd.tools.64.a(editor_translation_parser.linuxbsd.tools.64.o): In function `Ref<EditorTranslationParserPluginGDScript>::instance()':
/home/travis/build/godotengine/godot/./core/reference.h:228: undefined reference to `EditorTranslationParserPluginGDScript::EditorTranslationParserPluginGDScript()'
```